core/src/consensus/tendermint/vote_regression_checker.rs

```diff
@@ -0,0 +1,189 @@
+use consensus::{Step, VoteOn};
+use std::cmp::Ordering;
+
+pub struct VoteRegressionChecker {
+    last_vote: Option<VoteOn>,
+}
+
+impl VoteRegressionChecker {
+    pub fn new() -> VoteRegressionChecker {
+        VoteRegressionChecker {
+            last_vote: None,
+        }
+    }
+
+    pub fn check(&mut self, vote_on: &VoteOn) -> bool {
+        assert!(match vote_on.step.step {
+            Step::Propose | Step::Prevote | Step::Precommit => true,
+            _ => false,
+        });
+
+        let monotonic = if let Some(last_vote) = &self.last_vote {
+            match last_vote.step.cmp(&vote_on.step) {
+                Ordering::Less => true,
+                Ordering::Greater => false,
+                Ordering::Equal => last_vote.block_hash == vote_on.block_hash,
+            }
+        } else {
+            true
+        };
+
+        if monotonic {
+            self.last_vote = Some(vote_on.clone());
+        }
+        monotonic
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use consensus::VoteStep;
+    use primitives::H256;
+
+    #[test]
+    fn test_initial_set() {
+        let mut checker = VoteRegressionChecker::new();
+
+        let random_step = VoteStep::new(100, 10, Step::Prevote);
+        let random_hash = Some(H256::random());
+        assert!(checker.check(&VoteOn {
+            step: random_step,
+            block_hash: random_hash
+        }))
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Remove this for consistency.

Reply:
What does it mean 'consistency'?

Reply:
All the other tests in this PR are using `assert(!some_failing_condition)` instead of `#[should_panic]` + `assert(some_failing_condition)`.
Let's stick to one style.

Reply:
I've added a messgage at `assert!` on the code, and removed `assert!` in the test.

Reply:
I didn't get the point of your reply. Is the assertion message related here?

Reply:
The error is thrown by the `fn check()` isn't a recoverable error. The programmer shouldn't call it with `Step::Commit`. I won't make it return `false` when the user put `Step::Commit` in the `check`. In that case, This is the only way to remove `should_panic`:
```rust
std::panic::catch_unwind(|| {
    checker.check(&VoteOn {
        step: random_commit_step,
        block_hash: random_hash
    })
}
```
However I think this is somewhat ugly and I don't know why use this rather than `should_panic` to get 'consistency'. It was not the same kind of error at first place.

Reply:
I see. I misunderstood the code that the `check` function is returning `false` on `Commit` step.